### PR TITLE
checkInput function

### PR DESCRIPTION
**Pull request by jobo322**

### CODE REVIEW

File: src/checkInput.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+  let maxY = getMaxValue(data.y);
+  let y = new Array(x.length);
+  for (let i = 0; i < x.length; i++) {
+    y[i] = data.y[i] / maxY;
+  }
+
+  let {
+    minFactorWidth = 0.25,
+    maxFactorWidth = 4,
+    minFactorX = 2,
```
Comment (lpatiny) [inline src/checkInput.js:50]: Should this value not be 0.5 ?

File: src/index.js [hunk 2]
```diff
@@ -25,6 +21,18 @@ const keys = ['x', 'y', 'width', 'mu'];
  * @param {string} [options.shape.kind = 'gaussian'] - kind of shape; lorentzian, gaussian and pseudovoigt are supported.
  * @param {object} [options.optimization = {}] - it's specify the kind and options of the algorithm use to optimize parameters.
  * @param {object} [options.optimization.kind = 'lm'] - kind of algorithm. By default it's levenberg-marquardt.
+ * @param {number} [options.optimization.minFactorWidth = 0.25] - factor of width to define the lower limit of the width parameter.
+ * @param {number} [options.optimization.maxFactorWidth = 4] - factor of width to define the upper limit of the width parameter.
+ * @param {number} [options.optimization.minFactorX = 2] - factor of width to define the lower limit of the x parameter.
```
Comment (lpatiny) [inline src/index.js:26]: 0.5 ?

File: src/checkInput.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    throw new Error('kind should be a string');
+  }
+
+  let kind = shape.kind.toLowerCase().replace(/[^a-z]/g, '');
+
+  let nbParams;
+  let paramsFunc;
+  switch (kind) {
+    case 'gaussian':
+      nbParams = 3;
```
Comment (lpatiny) [inline src/checkInput.js:25]: We should I think make distinction between defaultParameters and the actual parameters that we will optimize based on the options.
Allowed parameters could also define the default ratio. 
```
defaultParameters: {
  x: {minFactor: 0.5, maxFactor. 2, gradientDifferent: xxx},
  y: {minFactor: 0, maxFactor: 1, gradientDifferent: xxx}},
  width: {minFactor: 0.5, maxFactor: 1, gradientDifferent: xxx}},
}
```

File: src/index.js [hunk 4]
```diff
@@ -25,6 +21,18 @@ const keys = ['x', 'y', 'width', 'mu'];
  * @param {string} [options.shape.kind = 'gaussian'] - kind of shape; lorentzian, gaussian and pseudovoigt are supported.
  * @param {object} [options.optimization = {}] - it's specify the kind and options of the algorithm use to optimize parameters.
  * @param {object} [options.optimization.kind = 'lm'] - kind of algorithm. By default it's levenberg-marquardt.
+ * @param {number} [options.optimization.minFactorWidth = 0.25] - factor of width to define the lower limit of the width parameter.
```
Comment (lpatiny) [inline src/index.js:24]:
```
options.optimization: {
x: {minFactor, maxFactor, gradientDifference},
y: {minFactor, maxFactor, gradientDifference},
}
```

**Review by lpatiny — COMMENTED** [review]
I would refactor the code to be more general and have one more level of hierarchy in the parameters.

File: src/checkInput.js (hunk 3 above)
Comment (jobo322) [reply src/checkInput.js:25]: We make gradientDifference for x and width depending of peak width, so I think it should have not a default value. I prefer your format

File: src/checkInput.js [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+          max: () => 1,
+          gradientDifference: () => 0.01,
+        },
+      };
+      break;
+    default:
+      throw new Error('kind of shape is not supported');
+  }
+
+  let x = data.x;
```
Comment (lpatiny) [inline src/checkInput.js:104]: Could you still check if we can get the same results by giving the right gradient value without rescaling. Would be nice not to have to make this rescaling.

File: src/checkInput.js [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+  let paramsFunc;
+  let defaultParameters;
+  switch (kind) {
+    case 'gaussian':
+      paramsFunc = sumOfGaussians;
+      defaultParameters = {
+        x: {
+          init: (peak) => peak.x,
+          max: (peak) => peak.x + peak.width * 2,
```
Comment (lpatiny) [inline src/checkInput.js:30]: Based on your experience we should really allow such a shift ? I would rather expect peak.width/2 but you may be right.

**Comment by lpatiny** [comment]
The code looks much more general and simple !
Please still update the jsdoc to reflect the changes in the parameters names.

File: src/checkInput.js (hunk 6 above)
Comment (jobo322) [reply src/checkInput.js:30]: When the guess is bad e.g the peak picking is failing due to noise data, it is good to have a little bit of freedom. it is the case for TGA data where the curve is not smooth and/or staggered.

File: src/checkInput.js (hunk 5 above)
Comment (jobo322) [reply src/checkInput.js:104]: I did try with maxY / 1e-3 which is equivalent to 1e-3 when data is rescaling dividing by maxY, and the result is not equal. I could search more.

**Pull request merged**